File: wrappers/python/iso3166/registry.py
"""
iso3166.registry — load and query the ISO 3166 country registry.

    from iso3166 import CountryRegistry

    reg = CountryRegistry()
    us = reg.active("US")
    if us is not None:
        print(us.alpha_3, us.name)

The registry file is bundled inside the package. An alternative path
can be passed to the constructor.
"""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Iterable, Iterator, Optional


__all__ = ["Country", "CountryRegistry", "RegistryError"]


ACTIVE_STATUSES = frozenset({
    "officially-assigned",
    "user-assigned",
    "exceptionally-reserved",
})


class RegistryError(Exception):
    """Raised when the registry file cannot be loaded."""


@dataclass(frozen=True)
class Country:
    """One country entry from the registry."""

    alpha_2: str
    alpha_3: str
    numeric: str
    name: str
    status: str
    independent: bool
    official_name: Optional[str] = None
    region: Optional[str] = None
    subregion: Optional[str] = None
    intermediate_region: Optional[str] = None
    currency_codes: Optional[list[str]] = None
    calling_codes: Optional[list[str]] = None
    tlds: Optional[list[str]] = None
    languages: Optional[list[str]] = None
    borders: Optional[list[str]] = None
    note: Optional[str] = None
    last_verified: Optional[str] = None
    withdrawal_date: Optional[str] = None
    replaced_by: Optional[list[str]] = None

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> "Country":
        return cls(
            alpha_2=d["alpha_2"],
            alpha_3=d["alpha_3"],
            numeric=d["numeric"],
            name=d["name"],
            status=d["status"],
            independent=d["independent"],
            official_name=d.get("official_name"),
            region=d.get("region"),
            subregion=d.get("subregion"),
            intermediate_region=d.get("intermediate_region"),
            currency_codes=d.get("currency_codes"),
            calling_codes=d.get("calling_codes"),
            tlds=d.get("tlds"),
            languages=d.get("languages"),
            borders=d.get("borders"),
            note=d.get("note"),
            last_verified=d.get("last_verified"),
            withdrawal_date=d.get("withdrawal_date"),
            replaced_by=d.get("replaced_by"),
        )

    @property
    def is_active(self) -> bool:
        return self.status in ACTIVE_STATUSES


def _default_registry_path() -> Path:
    return Path(__file__).resolve().parent / "iso3166.json"


class CountryRegistry:
    """Load and query the ISO 3166 registry.

    The registry is loaded once at construction. All lookups return
    Country instances (immutable); no method mutates internal state.
    """

    def __init__(self, path: str | Path | None = None) -> None:
        p = Path(path) if path is not None else _default_registry_path()
        if not p.exists():
            raise RegistryError(f"registry not found: {p}")
        try:
            data = json.loads(p.read_text(encoding="utf-8"))
        except json.JSONDecodeError as exc:
            raise RegistryError(f"{p}: invalid JSON: {exc}") from exc

        self._path = p
        self._meta: dict[str, Any] = data.get("meta", {})

        countries = data.get("countries", {})
        self._active: list[Country] = [
            Country.from_dict(e) for e in countries.get("active", [])
        ]
        self._withdrawn: list[Country] = [
            Country.from_dict(e) for e in countries.get("withdrawn", [])
        ]
        self._all: list[Country] = self._active + self._withdrawn

        # Indexes: preferred (active) entry per key.
        self._by_alpha2: dict[str, Country] = {}
        for c in self._active:
            self._by_alpha2[c.alpha_2] = c
        self._by_alpha3: dict[str, Country] = {c.alpha_3: c for c in self._active}
        self._by_numeric: dict[str, Country] = {c.numeric: c for c in self._active}

    # ------------------------------------------------------------
    # Metadata
    # ------------------------------------------------------------

    @property
    def path(self) -> Path:
        return self._path

    @property
    def meta(self) -> dict[str, Any]:
        return dict(self._meta)

    @property
    def version(self) -> str:
        return str(self._meta.get("version", ""))

    @property
    def updated(self) -> str:
        return str(self._meta.get("updated", ""))

    # ------------------------------------------------------------
    # Collections
    # ------------------------------------------------------------

    def all(self) -> list[Country]:
        """All entries, active first, then withdrawn, each sorted by alpha_2."""
        return list(self._all)

    def all_active(self) -> list[Country]:
        return list(self._active)

    def all_withdrawn(self) -> list[Country]:
        return list(self._withdrawn)

    def __iter__(self) -> Iterator[Country]:
        return iter(self._all)

    def __len__(self) -> int:
        return len(self._all)

    # ------------------------------------------------------------
    # Lookup
    # ------------------------------------------------------------

    def active(self, alpha_2: str) -> Optional[Country]:
        """Look up by alpha-2. Case-insensitive. Returns None if absent.

        Only active entries are returned; withdrawn codes are reachable
        via with_alpha2.
        """
        return self._by_alpha2.get(alpha_2.upper())

    def with_alpha2(self, alpha_2: str) -> list[Country]:
        """All entries (active and withdrawn) with this alpha-2 code.

        A small number of codes appear twice because ISO reassigned them
        (AI, SK). Callers who want a single entry should use active().
        """
        code = alpha_2.upper()
        return [c for c in self._all if c.alpha_2 == code]

    def by_alpha2(self, alpha_2: str) -> Optional[Country]:
        """Alias for active(). Kept for API symmetry with by_alpha3/by_numeric."""
        return self.active(alpha_2)

    def by_alpha3(self, alpha_3: str) -> Optional[Country]:
        return self._by_alpha3.get(alpha_3.upper())

    def by_numeric(self, numeric: str) -> Optional[Country]:
        return self._by_numeric.get(str(numeric).zfill(3))

    # ------------------------------------------------------------
    # Queries
    # ------------------------------------------------------------

    def currencies(self, alpha_2: str) -> list[str]:
        c = self.active(alpha_2)
        if c is None or not c.currency_codes:
            return []
        return list(c.currency_codes)

    def countries_with(self, currency_code: str) -> list[Country]:
        code = currency_code.upper()
        return [
            c for c in self._active
            if c.currency_codes and code in c.currency_codes
        ]

    def region(self, region_name: str) -> list[Country]:
        target = region_name.lower()
        return [
            c for c in self._active
            if (c.region or "").lower() == target
        ]

    def subregion(self, subregion_name: str) -> list[Country]:
        target = subregion_name.lower()
        return [
            c for c in self._active
            if (c.subregion or "").lower() == target
        ]

    def search(self, query: str) -> list[Country]:
        """Substring search on name, official_name, alpha_2, alpha_3."""
        q = query.lower()
        out: list[Country] = []
        for c in self._all:
            haystack = " ".join(filter(None, [
                c.name, c.official_name, c.alpha_2, c.alpha_3,
            ])).lower()
            if q in haystack:
                out.append(c)
        return out

    # ------------------------------------------------------------
    # Summary
    # ------------------------------------------------------------

    def summary(self) -> dict[str, Any]:
        by_status: dict[str, int] = {}
        for c in self._all:
            by_status[c.status] = by_status.get(c.status, 0) + 1
        return {
            "version": self.version,
            "updated": self.updated,
            "total": len(self._all),
            "active": len(self._active),
            "withdrawn": len(self._withdrawn),
            "by_status": by_status,
        }

File: wrappers/go/registry.go
// Package iso3166 loads and queries the ISO 3166 country registry.
//
// Usage:
//
//	reg, err := iso3166.Load()
//	if err != nil {
//	    log.Fatal(err)
//	}
//	if us := reg.Active("US"); us != nil {
//	    fmt.Println(us.Alpha3, us.Name)
//	}
//
// The registry file is bundled inside the package (iso3166.json).
// An alternative path can be passed to LoadFrom.
//
// Every lookup returns nil (or an empty slice) on a miss; only Load and
// LoadFrom return an error. This mirrors the Python and JavaScript
// wrappers, whose cross-language contract is checked against the same
// fixture (tests/cross_language_consistency.json).
package iso3166

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"os"
	"strings"
)

//go:embed iso3166.json
var bundledRegistry []byte

// Country is one entry from the registry. All pointer fields are nil
// when the registry entry has no value for them.
type Country struct {
	Alpha2             string   `json:"alpha_2"`
	Alpha3             string   `json:"alpha_3"`
	Numeric            string   `json:"numeric"`
	Name               string   `json:"name"`
	Status             string   `json:"status"`
	Independent        bool     `json:"independent"`
	OfficialName       *string  `json:"official_name"`
	Region             *string  `json:"region"`
	Subregion          *string  `json:"subregion"`
	IntermediateRegion *string  `json:"intermediate_region"`
	CurrencyCodes      []string `json:"currency_codes"`
	CallingCodes       []string `json:"calling_codes"`
	TLDs               []string `json:"tlds"`
	Languages          []string `json:"languages"`
	Borders            []string `json:"borders"`
	Note               *string  `json:"note"`
	LastVerified       *string  `json:"last_verified"`
	WithdrawalDate     *string  `json:"withdrawal_date"`
	ReplacedBy         []string `json:"replaced_by"`
}

// IsActive reports whether the entry is one of the three active statuses.
func (c *Country) IsActive() bool {
	switch c.Status {
	case "officially-assigned", "user-assigned", "exceptionally-reserved":
		return true
	}
	return false
}

// Meta mirrors the registry's meta block.
type Meta struct {
	Version        string `json:"version"`
	Updated        string `json:"updated"`
	Source         string `json:"source"`
	SchemaVersion  string `json:"schema_version"`
	CountActive    int    `json:"count_active"`
	CountWithdrawn int    `json:"count_withdrawn"`
}

// Summary is returned by CountryRegistry.Summary.
type Summary struct {
	Version   string         `json:"version"`
	Updated   string         `json:"updated"`
	Total     int            `json:"total"`
	Active    int            `json:"active"`
	Withdrawn int            `json:"withdrawn"`
	ByStatus  map[string]int `json:"by_status"`
}

type registryFile struct {
	Meta      Meta `json:"meta"`
	Countries struct {
		Active    []Country `json:"active"`
		Withdrawn []Country `json:"withdrawn"`
	} `json:"countries"`
}

// CountryRegistry is an immutable view of the registry. It is safe for
// concurrent reads.
type CountryRegistry struct {
	meta      Meta
	all       []*Country
	active    []*Country
	withdrawn []*Country

	byAlpha2  map[string]*Country
	byAlpha3  map[string]*Country
	byNumeric map[string]*Country

	path string
}

// Load reads the registry bundled inside the package.
func Load() (*CountryRegistry, error) {
	return LoadFrom("")
}

// LoadFrom reads the registry from path. An empty path uses the
// bundled copy.
func LoadFrom(path string) (*CountryRegistry, error) {
	var data []byte
	var err error

	if path == "" {
		data = bundledRegistry
	} else {
		data, err = os.ReadFile(path)
		if err != nil {
			return nil, fmt.Errorf("registry not found: %s: %w", path, err)
		}
	}

	var file registryFile
	if err := json.Unmarshal(data, &file); err != nil {
		return nil, fmt.Errorf("invalid registry JSON: %w", err)
	}

	r := &CountryRegistry{
		meta: file.Meta,
		path: path,
	}

	r.active = make([]*Country, len(file.Countries.Active))
	for i := range file.Countries.Active {
		r.active[i] = &file.Countries.Active[i]
	}
	r.withdrawn = make([]*Country, len(file.Countries.Withdrawn))
	for i := range file.Countries.Withdrawn {
		r.withdrawn[i] = &file.Countries.Withdrawn[i]
	}

	r.all = make([]*Country, 0, len(r.active)+len(r.withdrawn))
	r.all = append(r.all, r.active...)
	r.all = append(r.all, r.withdrawn...)

	r.byAlpha2 = make(map[string]*Country, len(r.active))
	for _, c := range r.active {
		r.byAlpha2[c.Alpha2] = c
	}
	r.byAlpha3 = make(map[string]*Country, len(r.active))
	for _, c := range r.active {
		r.byAlpha3[c.Alpha3] = c
	}
	r.byNumeric = make(map[string]*Country, len(r.active))
	for _, c := range r.active {
		r.byNumeric[c.Numeric] = c
	}

	return r, nil
}

// Meta returns a copy of the registry's meta block.
func (r *CountryRegistry) Meta() Meta { return r.meta }

// Version returns meta.version.
func (r *CountryRegistry) Version() string { return r.meta.Version }

// Updated returns meta.updated.
func (r *CountryRegistry) Updated() string { return r.meta.Updated }

// Path returns the path LoadFrom was given, or "" for the bundled copy.
func (r *CountryRegistry) Path() string { return r.path }

// All returns every entry (active and withdrawn).
func (r *CountryRegistry) All() []*Country {
	out := make([]*Country, len(r.all))
	copy(out, r.all)
	return out
}

// AllActive returns the active entries.
func (r *CountryRegistry) AllActive() []*Country {
	out := make([]*Country, len(r.active))
	copy(out, r.active)
	return out
}

// AllWithdrawn returns the withdrawn entries.
func (r *CountryRegistry) AllWithdrawn() []*Country {
	out := make([]*Country, len(r.withdrawn))
	copy(out, r.withdrawn)
	return out
}

// Active looks up an alpha-2 code. Case-insensitive. Returns nil if
// absent. Only active entries; withdrawn codes are reachable via
// WithAlpha2.
func (r *CountryRegistry) Active(alpha2 string) *Country {
	if alpha2 == "" {
		return nil
	}
	return r.byAlpha2[strings.ToUpper(alpha2)]
}

// WithAlpha2 returns all entries (active and withdrawn) with this
// alpha-2 code. A small number of codes appear twice because ISO
// reassigned them (AI, SK).
func (r *CountryRegistry) WithAlpha2(alpha2 string) []*Country {
	if alpha2 == "" {
		return nil
	}
	upper := strings.ToUpper(alpha2)
	var out []*Country
	for _, c := range r.all {
		if c.Alpha2 == upper {
			out = append(out, c)
		}
	}
	return out
}

// ByAlpha2 is an alias for Active.
func (r *CountryRegistry) ByAlpha2(alpha2 string) *Country {
	return r.Active(alpha2)
}

// ByAlpha3 looks up an alpha-3 code. Case-insensitive. Returns nil if
// absent.
func (r *CountryRegistry) ByAlpha3(alpha3 string) *Country {
	if alpha3 == "" {
		return nil
	}
	return r.byAlpha3[strings.ToUpper(alpha3)]
}

// ByNumeric looks up a numeric code. The argument is zero-padded to
// three digits, so "20" and "020" are equivalent.
func (r *CountryRegistry) ByNumeric(numeric string) *Country {
	if numeric == "" {
		return nil
	}
	padded := numeric
	for len(padded) < 3 {
		padded = "0" + padded
	}
	return r.byNumeric[padded]
}

// Currencies returns the currency codes in circulation in a country.
// Returns nil if the country is unknown or has none.
func (r *CountryRegistry) Currencies(alpha2 string) []string {
	c := r.Active(alpha2)
	if c == nil || c.CurrencyCodes == nil {
		return nil
	}
	out := make([]string, len(c.CurrencyCodes))
	copy(out, c.CurrencyCodes)
	return out
}

// CountriesWith returns the active countries where a currency
// circulates.
func (r *CountryRegistry) CountriesWith(currencyCode string) []*Country {
	if currencyCode == "" {
		return nil
	}
	target := strings.ToUpper(currencyCode)
	var out []*Country
	for _, c := range r.active {
		for _, cc := range c.CurrencyCodes {
			if cc == target {
				out = append(out, c)
				break
			}
		}
	}
	return out
}

// Region returns the active countries in a macro-region.
// Case-insensitive.
func (r *CountryRegistry) Region(name string) []*Country {
	if name == "" {
		return nil
	}
	target := strings.ToLower(name)
	var out []*Country
	for _, c := range r.active {
		if c.Region != nil && strings.ToLower(*c.Region) == target {
			out = append(out, c)
		}
	}
	return out
}

// Subregion returns the active countries in a sub-region.
// Case-insensitive.
func (r *CountryRegistry) Subregion(name string) []*Country {
	if name == "" {
		return nil
	}
	target := strings.ToLower(name)
	var out []*Country
	for _, c := range r.active {
		if c.Subregion != nil && strings.ToLower(*c.Subregion) == target {
			out = append(out, c)
		}
	}
	return out
}

// Search does a case-insensitive substring search on name,
// official_name, alpha_2, and alpha_3.
func (r *CountryRegistry) Search(query string) []*Country {
	if query == "" {
		return nil
	}
	q := strings.ToLower(query)
	var out []*Country
	for _, c := range r.all {
		haystack := c.Name
		if c.OfficialName != nil {
			haystack += " " + *c.OfficialName
		}
		haystack += " " + c.Alpha2 + " " + c.Alpha3
		if strings.Contains(strings.ToLower(haystack), q) {
			out = append(out, c)
		}
	}
	return out
}

// Summary returns counts per status and the total entry counts.
func (r *CountryRegistry) Summary() Summary {
	byStatus := make(map[string]int)
	for _, c := range r.all {
		byStatus[c.Status]++
	}
	return Summary{
		Version:   r.meta.Version,
		Updated:   r.meta.Updated,
		Total:     len(r.all),
		Active:    len(r.active),
		Withdrawn: len(r.withdrawn),
		ByStatus:  byStatus,
	}
}
